server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -1,90 +1,125 @@
+
+                  val attemptSet: F[Option[TokenAvailability]] = if (newTokenTotal >= 1) {
+                    setter((newTokenTotal - 1, currentTime))
+                      .map(_.guard[Option].as(TokenAvailable()))
+                  } else {
+                    val timeToFull = (refillEvery.toNanos * capacity) - timeDifference
+                    val successResponse = TokenUnavailable(timeToFull.nanos.some)
+                    setter((newTokenTotal, currentTime)).map(_.guard[Option].as(successResponse))
+                  }
+
+                  attemptSet
+                })
+            }
+
+            def loop: F[TokenAvailability] = attemptUpdate.flatMap { attempt =>
+              attempt.fold(loop)(token => token.pure[F])
+            }
+            loop
+          }
+
+        }
       }
-      Stream(bucket).concurrently(refill)
+
     }
-}
 
-/**
-  * Transform a service to reject any calls the go over a given rate with a 429 response.
-  */
-object Throttle {
+  }
 
   /**
     * Limits the supplied service to a given rate of calls using an in-memory [[TokenBucket]]
     *
     * @param amount the number of calls to the service to permit within the given time period.
     * @param per the time period over which a given number of calls is permitted.
     * @param http the service to transform.
-    * @return a singleton stream containing the transformed service.
+    * @return a task containing the transformed service.
     */
   def apply[F[_], G[_]](amount: Int, per: FiniteDuration)(
-      http: Http[F, G])(implicit F: Concurrent[F], timer: Timer[F]): Stream[F, Http[F, G]] = {
+      http: Http[F, G])(implicit F: Sync[F], timer: Clock[F]): F[Http[F, G]] = {
     val refillFrequency = per / amount.toLong
     val createBucket = TokenBucket.local(amount, refillFrequency)
     createBucket.map(bucket => apply(bucket)(http))
   }
 
+  def defaultResponse[F[_]](retryAfter: Option[FiniteDuration]): Response[F] = retryAfter match {
+    case Some(retryTime) => {
+      val retryHeader = `Retry-After`.unsafeFromDuration(retryTime)
+      Response[F](Status.TooManyRequests, headers = Headers(retryHeader))
+    }
+    case None => Response[F](Status.TooManyRequests)
+  }
+
   /**
     * Limits the supplied service using a provided [[TokenBucket]]
     *
     * @param bucket a [[TokenBucket]] to use to track the rate of incoming requests.
+    * @param throttleResponse a function that defines the response when throttled, may be supplied a suggested retry time depending on bucket implementation.
     * @param http the service to transform.
-    * @return a singleton stream containing the transformed service.
+    * @return a task containing the transformed service.
     */
-  def apply[F[_], G[_]](bucket: TokenBucket[F])(http: Http[F, G])(
-      implicit F: Concurrent[F]): Http[F, G] =
+  def apply[F[_], G[_]](
+      bucket: TokenBucket[F],
+      throttleResponse: Option[FiniteDuration] => Response[G] = defaultResponse[G] _)(
```

Comment:
I like this, because people can alternately fill in a timestamp instead of a duration if they prefer.  (Or of course add a custom body.)